File: py/rest_handler.py
import webapp2
from models import Game
from models import User
from google.appengine.ext import ndb
import logging
import json
from datetime import datetime
from google.appengine.api import users
from google.appengine.api import namespace_manager
from collections import defaultdict
import math
import datetime

class StatsHandler(webapp2.RequestHandler):
    def how_many_months_ago(self, date1, date2):
        delta = abs(date1 - date2)
        months = int(math.ceil(delta.days / 30))
        return months

    def initialize_stats(self):
        return {
            'count': 0,
            'nt_count': 0,
            'wins': 0,
            'win_per': 0.0,
            'history':[0,0,0,0,0,0,0,0,0,0,0,0]
        }

    def get(self):
        today = datetime.date.today()
        statMap = {}
        game_list = []
        namespace_manager.set_namespace(users.get_current_user().user_id())
        query = Game.query()
        games = query.fetch(1000)
        for game in games:
            faction = game.player_faction
            if not faction in statMap:
                statMap[faction] = self.initialize_stats()
            statMap[game.player_faction]['count'] += 1
            if game.won:
                statMap[game.player_faction]['wins'] += 1
            if not game.teaching and not game.draw:
                statMap[game.player_faction]['nt_count'] += 1

            months_ago = self.how_many_months_ago(game.date, today)
            if months_ago < 12:
                statMap[game.player_faction]['history'][months_ago] +=1


        self.response.out.write(json.dumps(statMap))


class GameHandler(webapp2.RequestHandler):

    def record_usage(self):
        cu = users.get_current_user()
        current_user = User.query().filter(User.user_id == cu.user_id()).get()
        if not current_user:
            user = User(user_id = cu.user_id(), name = cu.nickname(), email=cu.email())
            user.put()
        else:
            current_user.put()

    def get(self):
        """
        GET ALL games - and filters if there are search criteria
        """
        self.record_usage()
        game_list = []
        namespace_manager.set_namespace(users.get_current_user().user_id())
        query = Game.query()

        if self.request.get('player_faction'):
            query = query.filter(Game.player_faction == self.request.get('player_faction'))
        if self.request.get('player_warcaster'):
            query = query.filter(Game.player_warcaster == self.request.get('player_warcaster'))
        if self.request.get('opponent_name'):
           query = query.filter(Game.opponent_name == self.request.get('opponent_name'))
        if self.request.get('opponent_faction'):
            query = query.filter(Game.opponent_faction == self.request.get('opponent_faction'))
        if self.request.get('opponent_warcaster'):
            query = query.filter(Game.opponent_warcaster == self.request.get('opponent_warcaster'))
        if self.request.get('size'):
            query = query.filter(Game.size == self.request.get('size'))
        if self.request.get('result'):
            query = query.filter(Game.result == self.request.get('result'))
        if self.request.get('location'):
            query = query.filter(Game.location == self.request.get('location'))
        if self.request.get('game_type'):
            query = query.filter(Game.game_type == self.request.get('game_type'))

        win_count = 0
        non_teaching_count = 0

        games = query.fetch(1000)

        for game in games:
            if game.won:
                win_count = win_count + 1
            if not game.teaching and not game.draw:
                non_teaching_count = non_teaching_count + 1
            game_map = {}
            for attr in ['player_faction', 'player_warcaster', 'opponent_name', 'opponent_faction',
                         'opponent_warcaster', 'result', 'size', 'won', 'draw', 'teaching']:
                game_map[attr] = getattr(game, attr)
            game_map['date'] = game.date.isoformat()
            if game.created_at is not None:
                game_map['created_at'] = game.created_at.isoformat()
            game_map['key'] = game.key.urlsafe()
            game_list.append(game_map)
        game_list.sort(key=lambda game: game['date'], reverse=True)
        results = {
            'games': game_list,
            'win_count': win_count,
            'non_teaching_count': non_teaching_count
        }

        self.response.out.write(json.dumps(results))

    def post(self):
        """
        CREATE a new Game entity 
        """
        self.record_usage()
        namespace_manager.set_namespace(users.get_current_user().user_id())
        param_map = {}
        param_map['player_faction'] = self.request.get('player_faction')
        param_map['player_warcaster'] = self.request.get('player_warcaster')
        param_map['opponent_name'] = self.request.get('opponent_name', '')
        param_map['opponent_faction'] = self.request.get('opponent_faction')
        param_map['opponent_warcaster'] = self.request.get('opponent_warcaster')
        param_map['result'] = self.request.get('result')
        param_map['size'] = int(self.request.get('size', 0))
        param_map['won'] = self.request.get('won') == 'true'
        param_map['draw'] = self.request.get('draw') == 'true'
        param_map['teaching'] = self.request.get('teaching') == 'true'
        param_map['location'] = self.request.get('location', '')
        param_map['game_type'] = self.request.get('game_type', '')
        date = self.request.get('date')
        try:
            real_date = datetime.datetime.strptime(date,'%m/%d/%Y')
        except (ValueError):
            real_date = datetime.datetime.strptime(date,'%Y-%m-%d')
        param_map['date'] = real_date
        param_map['created_at'] = datetime.datetime.now()
        logging.error(param_map)
        game = Game(**param_map)
        game.put()
        self.response.out.write(game.key.urlsafe())

    def put(self):
        """
        UPDATE an existing Game entity
        """
        self.record_usage()
        namespace_manager.set_namespace(users.get_current_user().user_id())
        key_str = self.request.get('key')
        key = ndb.Key(urlsafe=key_str)
        game = key.get()
        if not game:
            self.response.set_code(500)
            return
        if self.request.get('player_faction'):
            game.player_faction = self.request.get('player_faction')
        if self.request.get('player_warcaster'):
            game.player_warcaster = self.request.get('player_warcaster')
        if self.request.get('opponent_name'):
            game.opponent_name = self.request.get('opponent_name')
        if self.request.get('opponent_faction'):
            game.opponent_faction = self.request.get('opponent_faction')
        if self.request.get('opponent_warcaster'):
            game.opponent_warcaster = self.request.get('opponent_warcaster')
        if self.request.get('result'):
            game.result = self.request.get('result')
        if self.request.get('size'):
            game.size = int(self.request.get('size', 0))
        if self.request.get('won'):
            game.won = bool(self.request.get('won'))
        if self.request.get('draw'):
            game.draw = bool(self.request.get('draw'))
        if self.request.get('teaching'):
            game.teaching = bool(self.request.get('teaching'))
        if self.request.get('date'):
            date = self.request.get('date')
            try:
                real_date = datetime.datetime.strptime(date,'%m/%d/%Y')
            except (ValueError):
                real_date = datetime.datetime.strptime(date,'%Y-%m-%d')
            game.date = real_date
        game.put()

    def delete(self):
        """
        DELETE an existing Game entity
        """
        self.record_usage()
        namespace_manager.set_namespace(users.get_current_user().user_id())
        key_str = self.request.url
        key_str = self.request.url[self.request.url.rfind('/')+1:]
        key = ndb.Key(urlsafe=key_str)
        key.delete()

app = webapp2.WSGIApplication([
    ('/game.*', GameHandler),
    ('/stat.*', StatsHandler)
], debug=True)

File: go/rest_handler.go
package fieldpromotion

import (
	"encoding/json"
	"net/http"
	"time"
	"sort"
	"strings"

	"appengine"
	"appengine/datastore"
	"appengine/user"
)

type Game struct {
	Date              time.Time      `datastore:"date" json:"date"`
	PlayerFaction     string         `datastore:"player_faction" json:"player_faction"`
	PlayerWarcaster   string         `datastore:"player_warcaster" json:"player_warcaster"`
	OpponentName      string         `datastore:"opponent_name" json:"opponent_name"`
	OpponentFaction   string         `datastore:"opponent_faction" json:"opponent_faction"`
	OpponentWarcaster string         `datastore:"opponent_warcaster" json:"opponent_warcaster"`
	Size              int            `datastore:"size" json:"size"`
	Result            string         `datastore:"result" json:"result"`
	Won               bool           `datastore:"won" json:"won"`
	Draw              bool           `datastore:"draw" json:"draw"`
	Teaching          bool           `datastore:"teaching" json:"teaching"`
	Location          string         `datastore:"location" json:"location"`
	GameType          string         `datastore:"game_type" json:"game_type"`
	Key               *datastore.Key `datastore:"-" json:"key"`
}

type ByDate []Game
func (a ByDate) Len() int           { return len(a) }
func (a ByDate) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByDate) Less(i, j int) bool { return !a[i].Date.Before(a[j].Date) }

type User struct {
	Email    string    `datastore:"email"`
	Name     string    `datastore:"name"`
	User_id  string    `datastore:"user_id"`
	Active   time.Time `datastore:"active"`
	Premeium bool      `datastore:"premeium"`
}

type Response struct {
	Games   []Game `json:"games"`
	Wins    int    `json:"win_count"`
	NtCount int    `json:"non_teaching_count"`
}

func handleGet(writer http.ResponseWriter, request *http.Request, namespace appengine.Context) {
	var games []Game
	var wins = 0
	var nts = 0

	query := datastore.NewQuery("Game")
	keys, _ := query.GetAll(namespace, &games)
	sort.Sort(ByDate(games))

	for i, key := range keys {
		games[i].Key = key
		if games[i].Won {
			wins++
		}

		if !games[i].Teaching && !games[i].Draw {
			nts++
		}
	}
	
	resp := Response{Games: games, Wins: wins, NtCount: nts}
	response, _ := json.Marshal(resp)
	writer.Write(response)
}

func handlePost(writer http.ResponseWriter, request *http.Request, namespace appengine.Context) {
	p := make([]byte, request.ContentLength)    
	_, err := request.Body.Read(p)

	if err == nil {
	    var newGame Game
	    err1 := json.Unmarshal(p, &newGame)
	    if err1 == nil {
	        key := datastore.NewKey(namespace, "Game", "", 0, nil)
			datastore.Put(namespace, key, &newGame)
	    } else {
	        namespace.Infof("Unable to unmarshall the JSON request", err1);
	    }
	}
}

func handleDelete(writer http.ResponseWriter, request *http.Request, namespace appengine.Context) {
	keyName := request.URL.Path
	i := strings.Index(keyName, "/")
	keyName = keyName[i+1:]
	i = strings.Index(keyName, "/")
	keyName = keyName[i+1:]
	
	key, err := datastore.DecodeKey(keyName)
	err = datastore.Delete(namespace, key)
	if err != nil {
		namespace.Infof("Yuck", err);
	}
}

func init() {
	http.HandleFunc("/game_go/", handleRequest)
}

func handleRequest(writer http.ResponseWriter, request *http.Request) {
	context := appengine.NewContext(request)
	user := user.Current(context)
	namespaceContext, _ := appengine.Namespace(context, user.ID)

	if request.Method == "GET" {
        handleGet(writer, request, namespaceContext)
    } else if request.Method == "POST" {
        handlePost(writer, request, namespaceContext)
    } else if request.Method == "DELETE" {
        handleDelete(writer, request, namespaceContext)
    }else {
        http.Error(writer, "Invalid request method.", 405)
    }
}
